lexicon/providers/njalla.py

```diff
@@ -0,0 +1,145 @@
+
+    def _authenticate(self):
+        params = {
+            'domain': self.domain,
+        }
+        result = self._api_call('get-domain', params)
+
+        if result['name'] != self.domain:
+            raise Exception('Domain not found')
+
+        self.domain_id = self.domain
+
+    # Create record. If record already exists with the same content, do nothing'
+    def _create_record(self, rtype, name, content):
+        params = {
+            'domain': self.domain,
+            'type': rtype,
+            'name': name,
+            'content': content,
+            'ttl': 10800,
+        }
+        if self._get_lexicon_option('ttl'):
+            params['ttl'] = self._get_lexicon_option('ttl')
+        result = self._api_call('add-record', params)
+
+        LOGGER.debug('create_record: %s', result)
+        return result
+
+    # List all records. Return an empty list if no records found
+    # type, name and content are used to filter records.
+    # If possible filter during the query, otherwise filter after response is received.
+    def _list_records(self, rtype=None, name=None, content=None):
+        params = {
+            'domain': self.domain,
+        }
+        result = self._api_call('list-records', params)
+
+        records = result['records']
+        processed_records = [{
+            'id': record['id'],
+            'type': record['type'],
+            'name': self._full_name(record['name']),
+            'ttl': record['ttl'],
+            'content': record['content'],
+        } for record in records]
+        filtered_records = [record for record in processed_records if (
+            (rtype is None or record['type'] == rtype) and
+            (name is None or record['name'] == self._full_name(name)) and
+            (content is None or record['content'] == content))]
+
+        LOGGER.debug('list_records: %s', filtered_records)
+        return filtered_records
+
+    # Create or update a record.
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        if not identifier:
+            identifier = self._get_record_identifier(rtype=rtype, name=name)
+
+        params = {
+            'id': identifier,
+            'domain': self.domain,
+            'content': content,
+        }
+        result = self._api_call('edit-record', params)
+
+        LOGGER.debug('update_record: %s', result)
+        return result
+
+    # Delete an existing record.
+    # If record does not exist, do nothing.
+    def _delete_record(self, identifier=None, rtype=None, name=None, content=None):
+        if not identifier:
+            identifier = self._get_record_identifier(rtype=rtype, name=name, content=content)
+
+        params = {
+            'domain': self.domain,
+            'id': identifier,
+        }
+        self._api_call('remove-record', params)
+
+        LOGGER.debug('delete_record: %s', True)
+        return True
+
+    # Helpers
+    def _api_call(self, method, params):
```

Comment:
Do you have a specific reason to not implement the method `_request` from the base provider? It seems to me that it would fit your case, and you could call `_post` in place of `_api_call`.

Reply:
I expected a comment on this. It just seemed weird to implement the `"request` method when the JSON-RPC API only takes POSTs. If I did, I'd still be tempted to wrap `_post` in another method that takes `method` and `params` args.

I just realized that the API docs are even require login, so some of that might not have been apparent. I'm willing to use `_request` if you feel that's important.

Reply:
I understand that for this provider specifically passing through `_request` may be a burden, but it is for a more general purpose. You see, Lexicon is composed of variety of providers developped by a lot of different people. At some point of Lexicon lifecycle, it is necessary to make some global modifications.

Having some "standard" APIs help a lot to avoid to miss something during these kind of refactoring. In particular, having the actual network requests passing through `_request` is helping to avoid missing specific logic if a global refactoring impact the network operations.

And I am completely ok to wrap the underlying `_request` calls to more suitable high level functions for the Njalla provider.

Reply:
I implemented `_request` :)